### PR TITLE
Port Commands to Pure Python

### PR DESCRIPTION
**Pull request by TheTripleV**

### CODE REVIEW

**Comment by auscompgeek** [comment]
Please create a separate PR with the tests.

File: commands2/__init__.py [hunk 1]
```diff
@@ -1,96 +1,87 @@
...
-    TrapezoidProfileSubsystemRadians,
-    Trigger,
-    WaitCommand,
-    WaitUntilCommand,
-    # button,
-    # cmd,
-    requirementsDisjoint,
+from . import button
+from .command import Command, InterruptionBehavior
+from .commandfactories import (
```
Comment (auscompgeek) [inline commands2/__init__.py:3]: Why have the command factories moved modules?

Comment (auscompgeek) [inline commands2/__init__.py:3]: Why have the command factories moved modules?

Comment (auscompgeek) [inline commands2/__init__.py:3]: Why have the command factories moved modules?

File: commands2/__init__.py [hunk 2]
```diff
@@ -1,96 +1,87 @@
...
+from .selectcommand import SelectCommand
+from .sequentialcommandgroup import SequentialCommandGroup
+from .startendcommand import StartEndCommand
+from .subsystem import Subsystem
+from .waitcommand import WaitCommand
+from .waituntilcommand import WaitUntilCommand
+from .wrappercommand import WrapperCommand
 
 __all__ = [
+    "button",
```
Comment (auscompgeek) [inline commands2/__init__.py:45]: Why is the button module re-exported?

File: commands2/__init__.py [hunk 3]
```diff
@@ -1,96 +1,87 @@
...
+    "print_",
+    "waitSeconds",
+    "waitUntil",
+    "either",
+    "select",
+    "sequence",
+    "repeatingSequence",
+    "parallel",
+    "race",
+    "deadline",
```
Comment (auscompgeek) [inline commands2/__init__.py:86]: Why are the command factories re-exported??

File: commands2/__init__.py [hunk 4]
```diff
@@ -1,96 +1,87 @@
...
+    "SelectCommand",
     "SequentialCommandGroup",
     "StartEndCommand",
     "Subsystem",
-    "SubsystemBase",
-    "Swerve2ControllerCommand",
-    "Swerve3ControllerCommand",
-    "Swerve4ControllerCommand",
-    "Swerve6ControllerCommand",
-    "TimedCommandRobot",
```
Comment (auscompgeek) [inline commands2/__init__.py:85]: What happened to TimedCommandRobot?

File: commands2/__init__.py (hunk 1 above)
Comment (TheTripleV) [reply commands2/__init__.py:3]: I organized it in a separate file as Java did. I didn't want to clutter `__init__.py` with them so I defined them in a separate file and reexported everything so you can still access them directly like in 2023.

File: commands2/__init__.py (hunk 4 above)
Comment (TheTripleV) [reply commands2/__init__.py:85]: Whoops forgot. Added it in. The java/cpp version isn't merged yet. https://github.com/wpilibsuite/allwpilib/pull/4871

File: commands2/__init__.py (hunk 1 above)
Comment (auscompgeek) [reply commands2/__init__.py]: My point is they were already in a module but you changed its name.

File: commands2/commandscheduler.py [hunk 5]
```diff
@@ -0,0 +1,471 @@
...
+                subsystem.simulationPeriodic()
+            # self._watchdog.addEpoch()
+
+        loopCache = self._activeButtonLoop
+        loopCache.poll()
+        self._watchdog.addEpoch("buttons.run()")
+
+        self._inRunLoop = True
+
+        for command in self._scheduledCommands.copy():
```
Comment (virtuald) [inline commands2/commandscheduler.py:219]: `_scheduledCommands` is a set, which doesn't have a stable iteration order. The Java version is a LinkedHashSet which does, and the C++ version uses a SmallSet which defaults to a vector which also does. Seems like this should also? (use a `dict[Command, None]`?)

File: commands2/parallelcommandgroup.py [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+
+        :param commands: the commands to include in this composition."""
+        super().__init__()
+        self._commands: Dict[Command, bool] = {}
+        self._runsWhenDisabled = True
+        self._interruptBehavior = InterruptionBehavior.kCancelIncoming
+        self.addCommands(*commands)
+
+    def addCommands(self, *commands: Command):
+        # breakpoint()
```
Comment (virtuald) [inline commands2/parallelcommandgroup.py:35]: remove

File: commands2/commandscheduler.py [hunk 7]
```diff
@@ -0,0 +1,471 @@
...
+        Whether the given commands are running. Note that this only works on commands that are directly
+        scheduled by the scheduler; it will not work on commands inside compositions, as the scheduler
+        does not see them.
+
+        :param commands: the command to query
+        :returns: whether the command is currently scheduled
+        """
+        return all(command in self._scheduledCommands for command in commands)
+
+    def requiring(self, subsystem: Subsystem) -> Union[None, Command]:
```
Comment (virtuald) [inline commands2/commandscheduler.py:366]: `Optional[Command]`?

File: commands2/command.py [hunk 8]
```diff
@@ -0,0 +1,545 @@
...
+
+    def getInterruptionBehavior(self) -> InterruptionBehavior:
+        """
+        How the command behaves when another command with a shared requirement is scheduled.
+
+        :returns: a variant of InterruptionBehavior, defaulting to {@link InterruptionBehavior#kCancelSelf kCancelSelf}.
+        """
+        return InterruptionBehavior.kCancelSelf
+
+    def withTimeout(self, seconds: float) -> Command:
```
[the remark this answers is not in the document]
Comment (auscompgeek) [reply commands2/command.py:170]: I believe both the Java and C++ interfaces hide the underlying types too.

[the remark this answers is not in the document]
Comment (virtuald) [reply commands2/command.py:170]: [Nope](https://github.com/wpilibsuite/allwpilib/blob/7496e0d2086567cbb08996a50e5a58ec97bb2b43/wpilibNewCommands/src/main/java/edu/wpi/first/wpilibj2/command/Command.java#L145).

File: commands2/commandscheduler.py [hunk 9]
```diff
@@ -0,0 +1,471 @@
...
+        """
+        CommandScheduler._instance = None
+
+    def __init__(self) -> None:
+        """
+        Gets the scheduler instance.
+        """
+        if CommandScheduler._instance is not None:
+            return
+        CommandScheduler._instance = self
```
[the remark this answers is not in the document]
Comment (TheTripleV) [reply commands2/commandscheduler.py:54]: Yeah, I can only get rid of it if I force the constructor to crash.

File: commands2/commandscheduler.py (hunk 5 above)
Comment (TheTripleV) [reply commands2/commandscheduler.py:219]: In Java, all of them are ordered so I moved more to dicts. It didn't affect the tests either way so thats good.

**Comment by lospugs** [comment]
I can't get inline commands to work though. 

Have a subsystem with the following function:
`    def displayGreenCommand(self) -> commands2.Command:
        return commands2.cmd.run(lambda: self.displayGreen(), [self])
`
        
 I. try binding it with 
 `
 self.driverController.B().whileTrue(self.led.displayGreenCommand())
 `
 
 When I try to run it I get the error:
 `
 AttributeError: module 'commands2' has no attribute 'cmd'
 `
 
 The __init__.py file has run in there, but don't know if my environment is even setup properly to know if this is a true bug, or I'm incomplete in my test bed.

**Comment by lospugs** [comment]
Ran my tests, things look good for me.  From an existing example we had to teach students in 2023, it's only four-ish sets of changes.  My workflow:
1- Checkout this repo
1- Create virtualenv
1- Run setup.py install (inside virtual env)
1- Verify sys.path is proper
1- Implement examples with both inlined and class definition files
1- Simulate code to test

Changes that are forced with this PR:
1- Change any reference from CommandBase to Command
1- Change any reference from SubsystemBase to Subsystem
1- Change any in-line method lambdas defined inside subsystem file from [self] as a list, to just self
1- Change button bindings in CommandXboxController from upper case button letters to lower case. (e.g. A() -> a() )

More verbose info in the CD thread:
https://www.chiefdelphi.com/t/robotpy-2024-beta/443117/11?u=newtoncrosby

**Comment by virtuald** [comment]
A version of this has been merged to main and pushed to pypi as version 2024.0.0b3. There are a lot of missing pieces, tracking those at https://github.com/robotpy/robotpy-commands-v2/issues/28

**Pull request closed without merging**